src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -0,0 +1,802 @@
+                                Marshal.PtrToStructure(structPtr, status);
+                                _dependentServices[i] = new ServiceController(_machineName, status);
+                            }
+                        }
+                        finally
+                        {
+                            Marshal.FreeHGlobal(enumBuffer);
+                        }
+                    }
+                    finally
+                    {
+                        Interop.mincore.CloseServiceHandle(serviceHandle);
+                    }
+                }
+
+                return _dependentServices;
+            }
+        }
+
+        /// The name of the machine on which this service resides.
+        public string MachineName
+        {
+            get
+            {
+                return _machineName;
+            }
+        }
+
+        /// Returns the short name of the service referenced by this object.
+        public string ServiceName
+        {
+            get
+            {
+                if (_name == null)
+                    GenerateNames();
+                return _name;
+            }
+        }
+
+        public unsafe ServiceController[] ServicesDependedOn
+        {
+            get
+            {
+                if (_servicesDependedOn != null)
+                    return _servicesDependedOn;
+
+                IntPtr serviceHandle = GetServiceHandle(Interop.SERVICE_QUERY_CONFIG);
+                try
+                {
+                    int bytesNeeded = 0;
+                    bool success = Interop.mincore.QueryServiceConfig(serviceHandle, IntPtr.Zero, 0, out bytesNeeded);
+                    if (success)
+                    {
+                        _servicesDependedOn = new ServiceController[0];
+                        return _servicesDependedOn;
+                    }
+
+                    int lastError = Marshal.GetLastWin32Error();
+                    if (lastError != Interop.ERROR_INSUFFICIENT_BUFFER)
+                        throw new Win32Exception(lastError);
+
+                    // get the info
+                    IntPtr bufPtr = Marshal.AllocHGlobal((IntPtr)bytesNeeded);
+                    try
+                    {
+                        success = Interop.mincore.QueryServiceConfig(serviceHandle, bufPtr, bytesNeeded, out bytesNeeded);
+                        if (!success)
+                            throw new Win32Exception(Marshal.GetLastWin32Error());
+
+                        Interop.QUERY_SERVICE_CONFIG config = new Interop.QUERY_SERVICE_CONFIG();
+                        Marshal.PtrToStructure(bufPtr, config);
+                        Dictionary<string, ServiceController> dependencyHash = new Dictionary<string, ServiceController>();
+
+                        char* dependencyChar = config.lpDependencies;
+                        if (dependencyChar != null)
```

Comment:
How common is the case where dependencyChar == null?  If it's common, you could avoid initializing dependencyHash above and only initialize it when there's actually something to be added to it.  And then below if dependencyHash is still null, you'd just return an empty array.